### PR TITLE
Implementation of directional GMF model

### PR DESCRIPTION
**Pull request by kwat0308**
This branch allows analysis for the new directional GMF model. This is already updated with the new composition loss model.

Again, I create a PR to ta_updates for now.

### CODE REVIEW

**Comment by kwat0308** [comment]
Same here, both branches (`add-ptypes` and `Nex-calc-gmf`) are merged as well. There is an issue with the checks due to the py3.8 tests.

**Comment by cescalara** [comment]
@kwat0308, does this PR need updating or are you waiting on my review?

**Comment by kwat0308** [comment]
Reviewing the commits, I think that this can be merged with `ta_updates`. So once you review it I think we can push this 👍

File: fancy/analysis/analysis.py [hunk 1]
```diff
@@ -612,7 +613,7 @@ def _simulate_deflections(self, kappas, Nrand=100, path_to_lens=None):
         This then returns the simulated arrival directions and detected energies
         at Earth after deflection.
 
-        Composition is accounted for by using the rigidities instead of energies in the lensing process.
+        Composition is accounted for by using thy instead of energies in the lensing process.
```
Comment (cescalara) [inline fancy/analysis/analysis.py:616]: typo?

File: fancy/analysis/analysis.py [hunk 2]
```diff
@@ -965,22 +966,23 @@ def _prepare_fit_inputs(self):
...
             self.fit_input["NthetaP"] = len(self.gmf_deflections._thetaP_grid)
             self.fit_input["thetaP_grid"] = np.rad2deg(self.gmf_deflections._thetaP_grid)
             self.fit_input["kappa_interp_grid"] = self.gmf_deflections._f_kappa(self.gmf_deflections._thetaP_grid)
 
             # add rigidity instead of energy
-            self.fit_input["Rdet"] = self.data.uhecr.energy / self.Zdet
-            self.fit_input["Rth"] = self.data.detector.Eth / self.Zdet
+            rigs = self.data.uhecr.rigidity if self.analysis_type == self.GMFdirection_type else self.data.uhecr.energy / self.Zdet
+            self.fit_input["Rdet"] = rigs
+            self.fit_input["Rth"] = self.data.detector.Eth / self.Zdet  # should modify this for directional model?
```
Comment (cescalara) [inline fancy/analysis/analysis.py:985]: Does this still need modifying as the comment suggested?

**Review by cescalara — APPROVED** [review]
Noticed a few commented out lines lying around that could probably also be cleaned up, but otherwise happy to merge

File: fancy/analysis/analysis.py (hunk 1 above)
Comment (kwat0308) [reply fancy/analysis/analysis.py:616]: Looks like a typo to me, I can fix this with a new commit

File: fancy/analysis/analysis.py (hunk 2 above)
Comment (kwat0308) [reply fancy/analysis/analysis.py:985]: I think it's not necessary, as initially I thought that the rigidity would vary as the UHECRs are backtracked which would modify the energy threshold. But we assume rigidity conservation so the same Rth at Earth will hold at the galactic boundary.

**Pull request merged**